src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -0,0 +1,1154 @@
+            ostringstream cut;
+            cut << "("
+                << word.str()
+                << " & "
+                << mask.str()
+                << ")"
+                << " >> "
+                << "(" << bits << "*(" << i << " % " << divisor << " ))";
+            // if it is signed, need to propagate sign... shift-up then down:
+            if (op->type.is_int())
+            {
+                rhs << "asint"
+                    << "("
+                    << cut.str()
+                    << " << "
+                    << (32 - bits)
+                    << ")"
+                    << " >> "
+                    << (32 - bits);
+            }
+            else
+            {
+                rhs << cut.str();
+            }
+        }
+#endif
+        print_assignment(op->type, rhs.str());
+        return;
+    }
+
+    // If we're loading a contiguous ramp, "unroll" the ramp into loads:
+    Expr ramp_base = is_ramp_one(op->index);
+    if (ramp_base.defined())
+    {
+        internal_assert(op->type.is_vector());
+
+        ostringstream rhs;
+        rhs << print_type(op->type)
+            << "(";
+        const int lanes = op->type.lanes();
+        for(int i = 0; i < lanes; ++i)
+        {
+            rhs << print_name(op->name)
+                << "["
+                << print_expr(ramp_base)
+                << "+" << i
+                << "]";
+            if (i < lanes-1)
+                rhs << ", ";
+        }
+        rhs << ")";
+
+        print_assignment(op->type, rhs.str());
+
+        return;
+    }
+
+    string id_index = print_expr(op->index);
+    //id_index = print_cast(op->type, op->index.type(), id_index);
+
+    // Get the rhs just for the cache.
+    bool type_cast_needed = !(allocations.contains(op->name) &&
+                              allocations.get(op->name).type == op->type);
+    //Type buffer_type = (op->image.defined()) ? op->image.type() : op->type;
+    //type_cast_needed |= (buffer_type != op->type);
+
+    ostringstream rhs;
+    if (type_cast_needed)
+    {
+        ostringstream element;
+        element << print_name(op->name)
+                << "[" << id_index << "]";
+        Type target_type = op->type;
+        Type source_type = allocations.get(op->name).type;
+        rhs << print_cast(target_type, source_type, element.str());
+    }
+    else
+    {
+        rhs << print_name(op->name);
+        rhs << "[" << id_index << "]";
+    }
+
+    std::map<string, string>::iterator cached = cache.find(rhs.str());
+    if (cached != cache.end()) {
+        id = cached->second;
+        return;
+    }
+
+    if (op->index.type().is_vector()) {
+        // If index is a vector, gather vector elements.
+        internal_assert(op->type.is_vector());
+
+        // This has to be underscore as print_name prepends and
+        // underscore to names without one and that results in a name
+        // mismatch of a Load appears as the value of a Let.
```

Comment:
This comment has some typos; I think you mean "This has to be underscore as print_name prepends an underscore to names without one and that results in a name mismatch if a Load appears as the value of a Let"

Reply:
This was a verbatim copy of code from `CodeGen_Metal_Dev.cpp`; I will go ahead and rectify the comments in that file too.